File: src/services/RazorpayService.ts
import { db } from '../config/firebase';
import { collection, addDoc, updateDoc, doc } from 'firebase/firestore';

interface OrderDetails {
  amount: number;
  planId: string;
  userId: string;
}

interface PaymentOptions {
  amount: number;
  currency: string;
  orderId: string;
  planId: string;
  userInfo: {
    name: string;
    email: string;
    contact: string;
  };
  onSuccess: (response: any) => void;
  onError: (error: any) => void;
}

declare global {
  interface Window {
    Razorpay: any;
  }
}

export const loadRazorpayScript = (): Promise<boolean> => {
  return new Promise((resolve) => {
    if (window.Razorpay) {
      resolve(true);
      return;
    }

    const script = document.createElement('script');
    script.src = 'https://checkout.razorpay.com/v1/checkout.js';
    script.onload = () => resolve(true);
    script.onerror = () => resolve(false);
    document.body.appendChild(script);
  });
};

export const createOrder = async (orderDetails: OrderDetails): Promise<{ orderId: string, razorpayOrderId: string }> => {
  try {
    // First create the order in Firestore
    const ordersRef = collection(db, 'orders');
    const orderDoc = await addDoc(ordersRef, {
      amount: orderDetails.amount,
      planId: orderDetails.planId,
      userId: orderDetails.userId,
      status: 'pending',
      createdAt: new Date(),
    });

    // Then create the Razorpay order through our server
    const response = await fetch('/api/create-razorpay-order', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        amount: orderDetails.amount * 100, // Convert to paise
        currency: 'INR',
        receipt: orderDoc.id,
      }),
    });

    if (!response.ok) {
      throw new Error('Failed to create Razorpay order');
    }

    const { razorpayOrderId } = await response.json();

    // Update the Firestore document with the Razorpay order ID
    await updateDoc(orderDoc, {
      razorpayOrderId,
    });

    return {
      orderId: orderDoc.id,
      razorpayOrderId,
    };
  } catch (error) {
    console.error('Error creating order:', error);
    throw error;
  }
};

export const initializeRazorpayPayment = async (options: PaymentOptions): Promise<void> => {
  const { amount, currency, orderId, planId, userInfo, onSuccess, onError } = options;

  const razorpayOptions = {
    key: process.env.REACT_APP_RAZORPAY_KEY_ID,
    amount: amount * 100, // Razorpay expects amount in paise
    currency,
    name: 'Triple A Gym',
    description: `Payment for ${planId} plan`,
    order_id: orderId,
    prefill: {
      name: userInfo.name,
      email: userInfo.email,
      contact: userInfo.contact,
    },
    handler: async function (response: any) {
      try {
        // Update order status in Firestore
        const orderRef = doc(db, 'orders', orderId);
        await updateDoc(orderRef, {
          status: 'completed',
          paymentId: response.razorpay_payment_id,
          signature: response.razorpay_signature,
          updatedAt: new Date(),
        });

        onSuccess(response);
      } catch (error) {
        console.error('Error updating order status:', error);
        onError(error);
      }
    },
    modal: {
      ondismiss: function () {
        onError(new Error('Payment cancelled by user'));
      },
    },
    theme: {
      color: '#10B981',
    },
  };

  const razorpay = new window.Razorpay(razorpayOptions);
  razorpay.open();
}; 

File: android/app/src/main/java/com/triplea/gymapp/services/RazorpayService.java
package com.triplea.gymapp.services;

import android.app.Activity;
import android.util.Log;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.FirebaseFirestore;
import com.razorpay.Checkout;
import com.razorpay.PaymentResultListener;

import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;

public class RazorpayService {
    private static final String TAG = "RazorpayService";
    private static final String RAZORPAY_KEY = "rzp_test_GEZQfBnCrf1uyR";
    
    private final Activity activity;
    private final FirebaseFirestore db;
    private final FirebaseAuth auth;
    private PaymentCallback callback;

    public interface PaymentCallback {
        void onPaymentSuccess(String paymentId, String orderId);
        void onPaymentError(int code, String message);
    }

    public RazorpayService(Activity activity) {
        this.activity = activity;
        this.db = FirebaseFirestore.getInstance();
        this.auth = FirebaseAuth.getInstance();
        Checkout.preload(activity.getApplicationContext());
    }

    public void startPayment(String planId, int amount, String planName, PaymentCallback callback) {
        this.callback = callback;
        
        try {
            Checkout checkout = new Checkout();
            checkout.setKeyID(RAZORPAY_KEY);

            JSONObject options = new JSONObject();
            options.put("name", "Triple A Gym");
            options.put("description", planName);
            options.put("currency", "INR");
            options.put("amount", amount * 100); // Razorpay expects amount in paise
            options.put("prefill.email", auth.getCurrentUser().getEmail());
            options.put("theme.color", "#10B981");

            // Create order in Firestore
            String orderId = createOrder(planId, amount);
            options.put("order_id", orderId);

            checkout.open(activity, options);
        } catch (Exception e) {
            Log.e(TAG, "Error starting payment: " + e.getMessage());
            if (callback != null) {
                callback.onPaymentError(-1, "Error starting payment: " + e.getMessage());
            }
        }
    }

    private String createOrder(String planId, int amount) {
        String orderId = db.collection("orders").document().getId();
        
        Map<String, Object> order = new HashMap<>();
        order.put("userId", auth.getCurrentUser().getUid());
        order.put("planId", planId);
        order.put("amount", amount);
        order.put("status", "created");
        order.put("createdAt", System.currentTimeMillis());
        
        db.collection("orders").document(orderId).set(order);
        
        return orderId;
    }

    public void handlePaymentSuccess(String paymentId, String orderId) {
        if (callback != null) {
            callback.onPaymentSuccess(paymentId, orderId);
        }

        // Update order status in Firestore
        Map<String, Object> updates = new HashMap<>();
        updates.put("status", "completed");
        updates.put("paymentId", paymentId);
        updates.put("completedAt", System.currentTimeMillis());
        
        db.collection("orders")
            .document(orderId)
            .update(updates)
            .addOnSuccessListener(aVoid -> Log.d(TAG, "Order updated successfully"))
            .addOnFailureListener(e -> Log.e(TAG, "Error updating order", e));
    }

    public void handlePaymentError(int code, String message) {
        if (callback != null) {
            callback.onPaymentError(code, message);
        }
    }
} 